parsec/core/backend_connection/transport.py

```diff
@@ -114,7 +127,280 @@ async def _connect(
+                target_url=url,
+                proxies=proxies,
+            )
+        # We don't handle multiple proxies so keep the first correct one and pray !
+        for proxy in proxies:
+            if proxy in ("DIRECT", ""):
+                # PAC explicitly told us not to use a proxy
+                return ""
+            elif proxy:
+                # Should be of style `PROXY 8.8.8.8:9999`
+                proxy_type, proxy_netloc = proxy.strip().split()
+                proxy_type = proxy_type.upper()
+                if proxy_type in ("PROXY", "HTTP"):
+                    return f"http://{proxy_netloc}"
+                elif proxy_type == "HTTPS":
+                    return f"https://{proxy_netloc}"
+                else:
+                    logger.warning(
+                        "Unsupported proxy type requested by .PAC proxy config",
+                        proxy=proxy,
+                        target_url=url,
+                    )
+
+        else:
+            return None
+
+    except Exception as exc:
+        logger.warning("Error while using .PAC proxy config", target_url=url, exc_info=exc)
+        return None
+
+
+def _get_proxy_from_os_config(hostname: str, use_ssl: bool) -> Optional[str]:
+    if __force_proxy_url:
+        logger.debug("Retreiving proxy from forced location", proxy_url=__force_proxy_url)
+        return __force_proxy_url
+    elif __force_proxy_pac_url:
+        # Don' try to superstep the forced config !
+        return None
+    else:
+        logger.debug("Retreiving proxy from system")
+
+    # Proxy config is accessed two times here: first to check proxy bypass,
+    # then to retrieve the proxy url. This is okay enough given proxy config
+    # is not supposed to change that often.
+
+    proxies = getproxies()
+    if proxy_bypass(hostname, proxies):
+        return None
+
+    proxy_type = "https" if use_ssl else "http"
+    proxy_url = proxies.get(proxy_type)
+    return proxy_url
+
+
+async def _maybe_connect_through_proxy(
+    hostname: str, port: int, use_ssl: bool
+) -> Optional[trio.abc.Stream]:
+    target_url = _build_http_url(hostname, port, use_ssl)
+
+    # First, try to get proxy from the infamous PAC config system
+    proxy_url = await trio.to_thread.run_sync(_get_proxy_from_pac_config, target_url, hostname)
+    if proxy_url:
+        logger.debug("Got proxy from .PAC config", proxy_url=proxy_url, target_url=target_url)
+    elif proxy_url == "":
+        logger.debug("Got .PAC config explictly specifying direct access", target_url=target_url)
+
+    # Fallback on direct proxy url config in environ variables & Windows registers table
+    if proxy_url is None:  # `proxy_url == ""` explicitly indicates no proxy should be use
+        proxy_url = _get_proxy_from_os_config(hostname, use_ssl)
+        if proxy_url:
+            logger.debug("Got proxy", proxy_url=proxy_url, target_url=target_url)
+
+    if proxy_url in (None, ""):
+        return None
```

Comment:
It feels like related things are too far apart. 

Shouldn't the following block from `_connect` replace the different `return None` statement?
```python
        logger.debug("Using direct (no proxy) connection", hostname=hostname, port=port)
        try:
            stream = await trio.open_tcp_stream(hostname, port)

        except OSError as exc:
            logger.debug(
                "Impossible to connect to backend", hostname=hostname, port=port, exc_info=exc
            )
            raise BackendNotAvailable(exc) from exc
```
This way the semantics of this function becomes `open_connection_using_proxy_config`.

Reply:
yup, all the proxy logic is now contained within `get_proxy` and `maybe_connect_through_proxy` (that internally uses `get_proxy` and wrap the given stream the same way adding ssl support works)

Reply:
What I meant is that the semantics of:
```python
async def maybe_connect_through_proxy(
    hostname: str, port: int, use_ssl: bool
) -> Optional[trio.abc.Stream]:
```
feels a bit weird and that the following would make more sense:
```python
async def connect_maybe_through_proxy(
    hostname: str, port: int, use_ssl: bool
) -> trio.abc.Stream:
```

Reply:
This is to mimic the wrapping system for SSL support.
On top of that I prefer having the proxy concepts hidden behind the `_connect` generic concept given it's a ugly detail we don't want to talk too much about 😄